deepwell/src/services/site_member/service.rs

```diff
@@ -0,0 +1,158 @@
+        );
+
+        let model = match Self::get_optional(ctx, SiteMembership { site_id, user_id })
+            .await?
+        {
+            // If membership is found, remove it by setting the leave date.
+            Some(member_model) => {
+                let mut model = member_model.into_active_model();
+                model.deleted_at = Set(Some(now()));
+                model.update(txn).await?
+            }
+
+            // If no membership is found, return BadRequest error.
+            None => {
+                tide::log::error!(
+                    "Could not remove user ID {user_id} from site ID {site_id} as they are not a member."
+                );
+                return Err(Error::BadRequest);
+            }
+        };
+
+        Ok(model)
+    }
+
+    #[inline]
+    pub async fn get(
+        ctx: &ServiceContext<'_>,
+        key: SiteMembership,
+    ) -> Result<SiteMemberModel> {
+        find_or_error(Self::get_optional(ctx, key)).await
+    }
+
+    /// Get whether a user is a member of a site or not. Returns `None` if no membership is found.
+    pub async fn get_optional(
+        ctx: &ServiceContext<'_>,
+        SiteMembership { site_id, user_id }: SiteMembership,
+    ) -> Result<Option<SiteMemberModel>> {
+        let txn = ctx.transaction();
+        let model = SiteMember::find()
+            .filter(
+                Condition::all()
+                    .add(site_member::Column::SiteId.eq(site_id))
+                    .add(site_member::Column::UserId.eq(user_id))
+                    .add(site_member::Column::DeletedAt.is_null()),
+            )
+            .one(txn)
+            .await?;
+        Ok(model)
+    }
+
+    /// Get all users of a site, ordered by oldest to newest.
+    pub async fn get_site_members(
+        ctx: &ServiceContext<'_>,
+        site_id: i64,
+    ) -> Result<Vec<SiteMemberModel>> {
+        let txn = ctx.transaction();
+
+        let models = SiteMember::find()
+            .filter(
+                Condition::all()
+                    .add(site_member::Column::SiteId.eq(site_id))
+                    .add(site_member::Column::DeletedAt.is_null()),
+            )
+            .order_by_asc(site_member::Column::MembershipId)
+            .all(txn)
+            .await?;
+
+        Ok(models)
+    }
+
+    /// Get all sites a user is a member of, ordered by oldest to newest.
+    pub async fn get_user_sites(
+        ctx: &ServiceContext<'_>,
+        user_id: i64,
+    ) -> Result<Vec<SiteMemberModel>> {
+        let txn = ctx.transaction();
+
+        let models = SiteMember::find()
+            .filter(
+                Condition::all()
+                    .add(site_member::Column::UserId.eq(user_id))
+                    .add(site_member::Column::DeletedAt.is_null()),
+            )
+            .order_by_asc(site_member::Column::MembershipId)
+            .all(txn)
+            .await?;
+
+        Ok(models)
+    }
```

Comment:
For a `get_membership_history()` method, it would be `user_id: i64`, `site_id: Option<i64>`, and then returning all the membership events in chronological order. Probably should have a TODO for pagination?